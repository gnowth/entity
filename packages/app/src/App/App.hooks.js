import _ from 'lodash';
import idx from 'idx';
import React from 'react';
import { DefaultProvider } from '@gnowth/default';

export default {
  useComponents(props, context) {
    return {
      DefaultProvider: props.defaults ? DefaultProvider : React.Fragment,
      IntlProvider: (props.intlProviderProps && context.intlProvider) || React.Fragment,
      ThemeProvider: (props.themeProviderProps && context.themeProvider) || React.Fragment,
    };
  },

  usePropsDefault(props, context, provider) {
    const defaults = React.useMemo(
      () => (
        _.isFunction(props.defaults)
          ? props.defaults(context.defaults)
          : { ...context.defaults, ...props.defaults }
      ),
      [context.defaults, props.defaults],
    );

    return provider === React.Fragment
      ? undefined
      : defaults;
  },

  usePropsIntl(props, context, provider) {
    const intlProviderProps = React.useMemo(
      () => (
        _.isFunction(props.intlProviderProps)
          ? props.intlProviderProps(context.intlProviderProps)
          : {
            ...context.intlProviderProps,
            ...props.intlProviderProps,
            defaultFormats: {
              ...idx(context, x => x.intlProviderProps.defaultFormats),
              ...idx(props, x => x.intlProviderProps.defaultFormats),
            },
            formats: {
              ...idx(context, x => x.intlProviderProps.formats),
              ...idx(props, x => x.intlProviderProps.formats),
            },
            messages: {
              ...idx(context, x => x.intlProviderProps.messages),
              ...idx(props, x => x.intlProviderProps.messages),
            },
          }
      ),
      [context.intlProviderProps, props.intlProviderProps],
    );

    return provider === React.Fragment
      ? undefined
      : intlProviderProps;
  },

  usePropsTheme(props, context, provider) {
    const theme = React.useMemo(
      () => (
        _.isFunction(props.themeProviderProps)
          ? props.themeProviderProps(context.themeProviderProps)
          : { ...context.themeProviderProps, ...props.themeProviderProps }
      ),
      [context.themeProviderProps, props.themeProviderProps],
    );

    return provider === React.Fragment
      ? undefined
      : theme;
  },
};
