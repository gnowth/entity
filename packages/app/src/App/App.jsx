import PropTypes from 'prop-types';
import React from 'react';

import { AppProvider, Context } from '../context';
import defaultHooks from './App.hooks';

function App(props) {
  const hooks = { ...defaultHooks, ...props.hooks };
  const context = React.useContext(Context);
  const { DefaultProvider, IntlProvider, ThemeProvider } = hooks.useComponents(props, context);

  const nextContext = {
    ...context,
    defaults: hooks.usePropsDefault(props, context, DefaultProvider),
    intlProviderProps: hooks.usePropsIntl(props, context, IntlProvider),
    themeProviderProps: hooks.usePropsTheme(props, context, ThemeProvider),
  };

  return (
    <AppProvider {...nextContext}>
      <IntlProvider {...nextContext.intlProviderProps}>
        <ThemeProvider {...nextContext.themeProviderProps}>
          <DefaultProvider {...nextContext.defaults}>
            { props.children }
          </DefaultProvider>
        </ThemeProvider>
      </IntlProvider>
    </AppProvider>
  );
}

App.propTypes = {
  children: PropTypes.node.isRequired,
  defaults: PropTypes.shape({}),
  hooks: PropTypes.exact({
    useComponents: PropTypes.func,
    usePropsDefault: PropTypes.func,
    usePropsIntl: PropTypes.func,
    usePropsTheme: PropTypes.func,
  }),
  intlProviderProps: PropTypes.exact({
    locale: PropTypes.string,
    messages: PropTypes.shape({}),
  }),
  themeProviderProps: PropTypes.exact({
    theme: PropTypes.shape({}),
  }),
};

App.defaultProps = {
  defaults: undefined,
  hooks: undefined,
  intlProviderProps: undefined,
  themeProviderProps: undefined,
};

export default React.memo(App);
