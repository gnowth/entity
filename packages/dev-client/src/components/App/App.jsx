import standardTheme from '@gnowth/theme-standard';
import PropTypes from 'prop-types';
import React from 'react';
import { EntityTitle } from '@entity/core';
import { AppRoot } from '@gnowth/app';
import { GlobalStyles } from '@gnowth/theme';
import { Provider } from 'react-redux';
import { IntlProvider } from '@private/react-intl';
import { ThemeProvider } from 'styled-components';

import * as theme from 'styles';

import defaults from '../../defaults';

const App = props => (
  <AppRoot
    defaults={defaults}
    entity={EntityTitle}
    intlProvider={IntlProvider}
    intlProviderProps={{ locale: 'en', messages: {} }}
    themeProvider={ThemeProvider}
    themeProviderProps={{ theme: { ...standardTheme, ...theme } }}
    {...props}
  >
    <Provider store={defaults.store}>
      { props.children }
    </Provider>

    <GlobalStyles />
  </AppRoot>
);

App.propTypes = {
  children: PropTypes.node.isRequired,
};

export default React.memo(App);
