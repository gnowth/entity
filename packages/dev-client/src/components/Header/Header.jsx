import PropTypes from 'prop-types';
import PropTypesPlus from '@gnowth/prop-types-plus';
import React from 'react';
import { useDefaultStyle } from '@gnowth/style';
import { UIBox, UIButton } from '@gnowth/ui';

import defaultLocales from './Header.locales';
import defaultStyles from './Header.styles';

const Header = (props) => {
  const locales = Object.assign({}, defaultLocales, props.locales);
  const styles = useDefaultStyle(defaultStyles, props.styles);

  return (
    <UIBox css={styles.header} justifyContent="flex-start">
      <UIButton
        css={styles.buttons}
        locale={locales.github}
        to="https://github.com/gnowth/react"
        variant="text"
      />

      <UIButton
        css={styles.buttons}
        locale={locales.style_guide}
        to="https://gnowth.github.io/react/style-guide"
        variant="text"
      />

      <UIButton
        css={styles.buttons}
        locale={locales.changelog}
        to="/pages/changelog"
        variant="text"
      />

      <UIButton
        css={styles.buttons}
        locale={locales.readme}
        to="/pages/readme"
        variant="text"
      />
    </UIBox>
  );
};

Header.propTypes = {
  locales: PropTypes.exact({
    changelog: PropTypesPlus.locale.isRequired,
    github: PropTypesPlus.locale.isRequired,
    readme: PropTypesPlus.locale.isRequired,
    style_guide: PropTypesPlus.locale.isRequired,
  }),
  styles: PropTypes.exact({
    buttons: PropTypesPlus.css,
    header: PropTypesPlus.css,
  }),
};

Header.defaultProps = {
  locales: undefined,
  styles: undefined,
};

export default React.memo(Header);
