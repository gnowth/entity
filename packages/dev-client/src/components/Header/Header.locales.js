import { defineMessages } from 'react-intl';

export default defineMessages({
  changelog: {
    defaultMessage: 'CHANGELOG',
    description: 'Link Changelog',
    id: 'component.header.changelog',
  },

  github: {
    defaultMessage: 'GitHub',
    description: 'Link GitHub',
    id: 'component.header.github',
  },

  readme: {
    defaultMessage: 'README',
    description: 'Link readme',
    id: 'component.header.readme',
  },

  style_guide: {
    defaultMessage: 'StyleGuide',
    description: 'Link style guide',
    id: 'component.header.style_guide',
  },
});
