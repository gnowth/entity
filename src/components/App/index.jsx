import PropTypes from 'prop-types';
import React from 'react';
import { AppRoot } from '@gnowth/app';
import { Provider } from 'react-redux';
import { IntlProvider } from 'react-intl';
import { ThemeProvider } from 'styled-components';

import * as theme from 'styles';
import GlobalStyles from 'styles/global';

import defaults from '../../defaults';

const App = props => (
  <AppRoot
    defaults={defaults}
    intlProvider={IntlProvider}
    intlProviderProps={{ locale: 'en', messages: {} }}
    themeProvider={ThemeProvider}
    themeProviderProps={{ theme }}
    {...props}
  >
    <Provider store={defaults.store}>
      { props.children }
    </Provider>

    <GlobalStyles />
  </AppRoot>
);

App.propTypes = {
  children: PropTypes.node.isRequired,
};

export default App;
