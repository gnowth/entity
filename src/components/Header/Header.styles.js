import { css } from 'styled-components';

export default {
  buttons: css`
    margin-right: 1rem;
  `,

  header: css`
    background: ${props => props.theme.vars.colorWhite || 'white'};
    padding: 1rem;

    ${props => props.theme.boxShadows.material1}
  `,
};
