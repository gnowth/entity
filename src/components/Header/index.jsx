import PropTypes from 'prop-types';
import PropTypesPlus from '@gnowth/prop-types-plus';
import React from 'react';
import { UIBox, UIButton } from '@gnowth/ui';

import locales from './locales';
import styles from './styles';

const Header = props => (
  <UIBox css={props.styles.header} justifyContent="flex-start">
    <UIButton
      css={props.styles.buttons}
      locale={props.locales.github}
      to="https://github.com/gnowth/react"
      variant="text"
    />

    <UIButton
      css={props.styles.buttons}
      locale={props.locales.style_guide}
      to="https://gnowth.github.io/react/style-guide"
      variant="text"
    />

    <UIButton
      css={props.styles.buttons}
      locale={props.locales.changelog}
      to="/pages/changelog"
      variant="text"
    />

    <UIButton
      css={props.styles.buttons}
      locale={props.locales.readme}
      to="/pages/readme"
      variant="text"
    />
  </UIBox>
);

Header.propTypes = {
  locales: PropTypes.exact({
    changelog: PropTypesPlus.locale.isRequired,
    github: PropTypesPlus.locale.isRequired,
    readme: PropTypesPlus.locale.isRequired,
    style_guide: PropTypesPlus.locale.isRequired,
  }),
  styles: PropTypes.exact({
    buttons: PropTypesPlus.css,
    header: PropTypesPlus.css,
  }),
};

Header.defaultProps = {
  locales,
  styles,
};

export default Header;
