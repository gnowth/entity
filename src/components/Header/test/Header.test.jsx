import renderer from 'react-test-renderer';
import React from 'react';
import { HashRouter as Router } from 'react-router-dom';

import App from 'components/App';

import Header from '..';

jest.mock('store');

it('renders correctly', () => {
  const tree = renderer.create(
    <App>
      <Router>
        <Header />
      </Router>
    </App>,
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
