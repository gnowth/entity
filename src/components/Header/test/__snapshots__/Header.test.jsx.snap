// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="Box__UIBox-j97zbg-0 sc-dnqmqq jEPJNE"
>
  <a
    className="sc-iwsKbI gfniVm Buttonstyles__Button-mk3f56-0-a choPNg"
    href="https://github.com/gnowth/react"
    to="https://github.com/gnowth/react"
  >
    GitHub
  </a>
  <a
    className="sc-gZMcBi iDaNrw Buttonstyles__Button-mk3f56-0-a choPNg"
    href="https://gnowth.github.io/react/style-guide"
    to="https://gnowth.github.io/react/style-guide"
  >
    StyleGuide
  </a>
  <a
    _css4={
      Array [
        "margin-right:1rem;",
      ]
    }
    className="sc-gqjmRU lpbkHd Buttonstyles__Button-mk3f56-0-Link fwYiwn"
    href="#/pages/changelog"
    locale={
      Object {
        "defaultMessage": "CHANGELOG",
        "description": "Link Changelog",
        "id": "component.header.changelog",
      }
    }
    onClick={[Function]}
    variant="text"
  >
    CHANGELOG
  </a>
  <a
    _css5={
      Array [
        "margin-right:1rem;",
      ]
    }
    className="sc-VigVT ibkFLQ Buttonstyles__Button-mk3f56-0-Link fwYiwn"
    href="#/pages/readme"
    locale={
      Object {
        "defaultMessage": "README",
        "description": "Link readme",
        "id": "component.header.readme",
      }
    }
    onClick={[Function]}
    variant="text"
  >
    README
  </a>
</div>
`;
